xonsh/foreign_shells.py

```diff
@@ -139,6 +139,8 @@ def foreign_shell_data(shell, interactive=True, login=False, envcmd='env',
         currenv = dict(currenv)
     try:
         s = subprocess.check_output(cmd, stderr=subprocess.PIPE, env=currenv,
+                                    # setsid where available to avoid hangs
+                                    preexec_fn=getattr(os, 'setsid', None),
```

Comment:
Hi @minrk, thanks for putting this in.  I am a little hesitant to merge this in as-in due to the warnings in the [Python docs](https://docs.python.org/3/library/subprocess.html#popen-constructor) namely: 

```
Warning The preexec_fn parameter is not safe to use in the presence of threads in your application. 
The child process could deadlock before exec is called. If you must use it, keep it trivial! 
Minimize the number of libraries you call into.

Note If you need to modify the environment for the child use the env parameter rather than doing it in a preexec_fn.
The start_new_session parameter can take the place of a previously common use of preexec_fn to call os.setsid() 
in the child.
```

How do you feel about `start_new_session=True` that @takluyver suggested in #624?  This seems to be the safe version of what is implemented here.